File: 2023/03.03/정호조/이중우선순위큐.java
import java.util.Collections;
import java.util.PriorityQueue;

public class 이중우선순위큐 {
    public static void main(String[] args) {
        String[] operations = {"I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333"};
        Solution so = new Solution();
        System.out.println(so.solution(operations)[0] + " , " + so.solution(operations)[1]);
    }

    static class Solution {
        public int[] solution(String[] operations) {
            PriorityQueue<Integer> minHeap = new PriorityQueue<>();
            PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

            for (int i = 0; i < operations.length; i++) {
                if(operations[i].charAt(0) == 'I'){
                    minHeap.add(Integer.parseInt(operations[i].substring(2)));
                    maxHeap.add(Integer.parseInt(operations[i].substring(2)));
                } else if(operations[i].charAt(0) == 'D'){
                    if(maxHeap.size() > 0){
                        if(operations[i].charAt(2) == '1'){
                            int max = maxHeap.peek();
                            maxHeap.remove(max);
                            minHeap.remove(max);
                        } else {
                            int min = minHeap.peek();
                            maxHeap.remove(min);
                            minHeap.remove(min);
                        }
                    }
                }
            }
            int[] answer = new int[2];

            if(maxHeap.size() > 0){
                answer[0] = maxHeap.peek();
                answer[1] = minHeap.peek();
            } else {
                answer[0] = 0;
                answer[1] = 0;
            }

            return answer;
        }
    }
}

File: 2023/03.03/조수연/이중우선순위큐.kt
import java.util.*

fun main() {
    val operations = arrayOf("I -45", "I 653", "D 1", "I -642", "I 45", "I 97", "D 1", "D -1", "I 333")
    Solution().solution(operations)
}

class Solution {
    fun solution(operations: Array<String>): IntArray {
        var answer: IntArray
        val minPq = PriorityQueue<Int>() //오름차순
        val maxPq = PriorityQueue<Int>(Collections.reverseOrder()) //내림차순
        for (operation in operations) {
            val command = operation.split(" ")[0] //삽입, 삭제 명령어
            val data = operation.split(" ")[1].toInt()

            when (command) {
                "I" -> {
                    minPq.offer(data)
                    maxPq.offer(data)
                }
                "D" -> {
                    when (data) {
                        1 -> {
                            if (maxPq.isNotEmpty()) {
                                minPq.remove(maxPq.peek()) //minPq에서 maxPq.peek 부분도 같이 제거해주기
                                maxPq.poll()
                            }
                        }
                        -1 -> {
                            if (minPq.isNotEmpty()) {
                                maxPq.remove(minPq.peek()) //maxPq에서 minPq.peek 부분도 같이 제거해주기
                                minPq.poll()
                            }
                        }
                    }
                }
            }
        }
        if (maxPq.isEmpty() && minPq.isEmpty()) answer = intArrayOf(0, 0) //큐가 비어있으면 0,0
        else answer = intArrayOf(maxPq.peek(), minPq.peek())
        return answer
    }
}